Replace `external_trigger` check with DagRunType

jason810496 opened the pull request:
closes: #45932 


<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

ashb commented on airflow/example_dags/plugins/event_listener.py:
```diff
@@ -147,9 +147,8 @@ def on_dag_run_failed(dag_run: DagRun, msg: str):
-    external_trigger = dag_run.external_trigger
 
-    print(f"Dag information:{dag_id} Run id: {run_id} external trigger: {external_trigger}")
+    print(f"Dag information:{dag_id} Run id: {run_id}")
```
Not that it _matters_, but lets include the run_type here in its place.

ashb commented on airflow/jobs/scheduler_job_runner.py:
```diff
@@ -1476,14 +1476,18 @@ def _update_state(dag: DAG, dag_run: DagRun):
-            if dag.timetable.periodic and not dag_run.external_trigger and dag_run.clear_number < 1:
+            if (
+                dag.timetable.periodic
+                and not dag_run.run_type == DagRunType.MANUAL
```
```suggestion
                and dag_run.run_type != DagRunType.MANUAL
```

ashb commented on airflow/migrations/versions/0056_3_0_0_remove_external_trigger_field.py:
```diff
@@ -0,0 +1,56 @@
+
+def upgrade():
+    """Apply remove external_trigger field."""
+    # ### commands auto generated by Alembic - please adjust! ###
```
```suggestion
```

ashb commented on airflow/migrations/versions/0056_3_0_0_remove_external_trigger_field.py:
```diff
@@ -0,0 +1,56 @@
+
+def downgrade():
+    """Unapply remove external_trigger field."""
+    # ### commands auto generated by Alembic - please adjust! ###
```
```suggestion
```

ashb commented on airflow/migrations/versions/0056_3_0_0_remove_external_trigger_field.py:
```diff
@@ -0,0 +1,56 @@
+    with op.batch_alter_table("dag_run", schema=None) as batch_op:
+        batch_op.add_column(sa.Column("external_trigger", sa.BOOLEAN(), autoincrement=False, nullable=True))
+
+    # ### end Alembic commands ###
```
```suggestion
```

ashb commented on airflow/migrations/versions/0056_3_0_0_remove_external_trigger_field.py:
```diff
@@ -0,0 +1,56 @@
+    with op.batch_alter_table("dag_run", schema=None) as batch_op:
+        batch_op.drop_column("external_trigger")
+
+    # ### end Alembic commands ###
```
```suggestion
```

ashb commented on airflow/models/dagrun.py:
```diff
@@ -1227,11 +1217,11 @@ def _emit_true_scheduling_delay_stats_for_finished_state(self, finished_tis: lis
 
         Note that the stat will only be emitted for scheduler-triggered DAG runs
-        (i.e. when ``external_trigger`` is *False* and ``clear_number`` is equal to 0).
+        (i.e. when ``run_type`` is *MANUAL* and ``clear_number`` is equal to 0).
```
```suggestion
        (i.e. when ``run_type`` is not *MANUAL* and ``clear_number`` is equal to 0).
```

Though this brings up another question: How should this function deal with run_type==BACKFILL?

jason810496 replied:
How about adding an externally_triggered_type in airflow.utils.types?  Instead of checking `run_type == DagRunType.MANUAL`, we could use `run_type in externally_triggered_type`. WDYT?
```python
externally_triggered_type: frozenset[DagRunType] = frozenset([DagRunType.BACKFILL_JOB, DagRunType.MANUAL, DagRunType.ASSET_TRIGGERED])
```

ashb replied:
I think we need to look at what the intent here was, and how that maps on to the new dag run types. My guess is that when "external_trigger" was added, there was either scheduled, or external.

So Manual is clearly not scheduled, but what abouty Backfill runs? How should we treat those. And for that I delegate to @uranusjr and @dstandish *boing flip* 😄

jason810496 replied:
After looking deeper, I think `external_trigger is False` is equivalent to `run_type is SCHEDULED` and should fit this case.

uranusjr replied:
I dug into the history and yes it seems that `external_trigger` is *False* when `run_type` is set to `BACKFILL_JOB`. Also, `external_trigger` predates both `run_type` and `triggered_by` by a lot, its true definition is quite ambiguous and lost to time, so we can probably take some liberty changing its definition as long as the new implementation makes sense and does not cause bugs.

I think it’d be reasonable enough to define `external_trigger == True` as equivalent to `run_type == "manual"` for refactoring purposes. If this causes bugs, they should be fixed under this definition.

ashb commented on airflow/utils/db_cleanup.py:
```diff
@@ -107,9 +107,9 @@ def readable_config(self):
+        extra_columns=["dag_id", "run_type"],
         keep_last=True,
-        keep_last_filters=[column("external_trigger") == false()],
+        keep_last_filters=[column("run_type") != "manual"],
```
We shouldn't use a string literal here but the enum constant please.

ashb commented on providers/src/airflow/providers/standard/operators/latest_only.py:
```diff
@@ -53,7 +54,7 @@ def choose_branch(self, context: Context) -> str | Iterable[str]:
         dag_run: DagRun = context["dag_run"]  # type: ignore[assignment]
-        if dag_run.external_trigger:
+        if dag_run.run_type == DagRunType.MANUAL:
             self.log.info("Externally triggered DAG_Run: allowing execution to proceed.")
```
This message is wrong now

ashb commented on providers/src/airflow/providers/standard/operators/latest_only.py:
```diff
@@ -53,7 +54,7 @@ def choose_branch(self, context: Context) -> str | Iterable[str]:
         # skipping downstream tasks
         dag_run: DagRun = context["dag_run"]  # type: ignore[assignment]
-        if dag_run.external_trigger:
+        if dag_run.run_type == DagRunType.MANUAL:
```
Ditto here about BACKFILL. Need to decide how to handle this

dstandish replied:
so this is about short circuiting the "latest only check" when dag is externally-triggered.  I don't think it makes a lot of sense to short circuit this when backfilling since, isn't that the point of the latest only operator?

ashb commented on providers/tests/openlineage/plugins/test_adapter.py:
```diff
@@ -585,9 +586,8 @@ def test_emit_dag_started_event(mock_stats_incr, mock_stats_timer, generate_stat
-            "external_trigger": False if AIRFLOW_V_3_0_PLUS else None,
             "run_id": run_id,
-            "run_type": None,
+            "run_type": DagRunType.MANUAL if AIRFLOW_V_3_0_PLUS else None,
```
```suggestion
            "run_type": DagRunType.MANUAL,
```

run_type has been around for _a long time_, we don't need this switch I don't think?

ashb reviewed (commented):
This also needs a newsfragment noting about the removal of it please

jason810496 committed:
Replace external_trigger check with DagRunType

- Fix test_handle_multiple_columns_unique_constraint_error

- Resolve ashb code review

- Add newsfragment

- Replace external_trigger is False logic with run_type is SCHEDULED

- Fix _emit_true_scheduling_delay_stats_for_finished_state

- Fix migrations fix after rebasing to latest main

Fix test_exceptions and static check after rebasing

Fix test_dag_run and erd static check

uranusjr commented on newsfragments/45961.significant.rst:
```diff
@@ -0,0 +1,12 @@
+Replace ``external_trigger`` check with ``DagRunType``
```
We should mention if the user currently relies on `external_trigger`, they can replace it by checking `run_type`.

uranusjr commented on airflow/migrations/versions/0059_3_0_0_remove_external_trigger_field.py:
```diff
@@ -0,0 +1,49 @@
+def downgrade():
+    """Unapply remove external_trigger field."""
+    with op.batch_alter_table("dag_run", schema=None) as batch_op:
+        batch_op.add_column(sa.Column("external_trigger", sa.BOOLEAN(), autoincrement=False, nullable=True))
```
Downgrade should try to recreate the original external_trigger value using run_type.

uranusjr commented on airflow/utils/types.py:
```diff
@@ -57,6 +57,11 @@ def from_run_id(run_id: str) -> DagRunType:
 
+externally_triggered_type: frozenset[DagRunType] = frozenset(
+    [DagRunType.BACKFILL_JOB, DagRunType.MANUAL, DagRunType.ASSET_TRIGGERED]
+)
```
Where is this used? (Also the name should be plural.)

jason810496 replied:
This is not currently used, I will remove it. It came up during the discussion above, but instead of introducing `externally_triggered_type`, simply comparing `run_type != DagRunType.SCHEDULED` is a much cleaner approach.

uranusjr commented on providers/standard/src/airflow/providers/standard/operators/latest_only.py:
```diff
@@ -53,8 +54,8 @@ def choose_branch(self, context: Context) -> str | Iterable[str]:
         dag_run: DagRun = context["dag_run"]  # type: ignore[assignment]
-        if dag_run.external_trigger:
-            self.log.info("Externally triggered DAG_Run: allowing execution to proceed.")
+        if dag_run.run_type != DagRunType.SCHEDULED:
```
Should this check against MANUAL instead?

jason810496 replied:
Thanks for the reminder, just resolved.

uranusjr commented on newsfragments/45961.significant.rst:
I think this change should be covered by AIR302? cc @Lee-W

Lee-W replied:
Yep, I think so. It's not something that has been moved to providers or missing args. AIR302 should be able to be covered. Would be even better if @jason810496 can provide us with an example so that I can have a better idea on how this should be migrated

uranusjr replied:
I think the only user-facing interface Ruff can cover are

1. DAG.create_dagrun. You used to be able to pass in `external_trigger=True` (or False) and that’s set on the created DagRun row. This is now removed.
2. When accessing a DagRun (in context; I don’t think there are other officially supported ways?) there used to be an attribute `external_trigger` that’s now gone. `external_trigger=True` basically means `run_type="manual"`.

There are also REST API changes but I don’t think we have a way to detect those.

Lee-W replied:
Yep, both should be easy to add. We decided we don't want to handle REST API as there are way too many ways to interact with it

uranusjr commented on providers/openlineage/tests/unit/openlineage/plugins/test_adapter.py:
```diff
@@ -605,9 +606,8 @@ def test_emit_dag_started_event(mock_stats_incr, mock_stats_timer, generate_stat
-            "external_trigger": False if AIRFLOW_V_3_0_PLUS else None,
             "run_id": run_id,
-            "run_type": None,
+            "run_type": DagRunType.MANUAL if AIRFLOW_V_3_0_PLUS else None,
```
Does this run_type need to be different from different versions? Especially since external_trigger is False, should this always be SCHEDULED instead?

jason810496 replied:
I'm not really sure about _will run_type will be when external_trigger is False_ , but based on https://github.com/apache/airflow/pull/45961#discussion_r1928395914 we don't need to check airflow version when asserting run_type.

jason810496 committed:
fixup! Fix run_type logic, remove unused externally_triggered_type

jason810496 committed:
fixup! Restore external_trigger value when downgrade

uranusjr committed:
Remove externally_trigger reference in frontend

The 'triggered_by' field used to only be shown when the run is
externally triggered. Rather than using run_type to maintain the
conditional, I decided to render the field unconditionally instead,
since the value could be useful for scheduler-created runs too (for
example, distinguishing whether a run is asset-triggered or
time-scheduled).

Lee-W commented on newsfragments/45961.significant.rst:
```diff
@@ -0,0 +1,13 @@
+  * [ ] Behaviour changes
+  * [ ] Plugin changes
+  * [ ] Dependency changes
+  * [x] Code interface changes
```
Oops, missing migration rules here.

jason810496 replied:
> Oops, missing migration rules here.

Hi @Lee-W, do you mean we should add example for migrating `external_trigger` with `DagRunType` ? I have run `check-significant-newsfragments-are-valid` pre-commit successfully.

Lee-W replied:
just updated it. you can take a look at what I have added

jason810496 replied:
Thanks! I just took a look at the commit.

Lee-W commented on newsfragments/45961.significant.rst:
```diff
@@ -0,0 +1,13 @@
+
+  * [ ] Dag changes
+  * [ ] Config changes
+  * [ ] API changes
```
It also contains API and CLI changes

Lee-W committed:
docs(newsfragments): add migration rules

ashb commented on airflow/api_fastapi/execution_api/datamodels/taskinstance.py:
```diff
@@ -230,7 +230,6 @@ class DagRun(StrictBaseModel):
     clear_number: int
     run_type: DagRunType
     conf: Annotated[dict[str, Any], Field(default_factory=dict)]
-    external_trigger: bool = False
```
This will need a comparable change made into `task_sdk/src/airflow/sdk/api/datamodels/_generated.py` (It's not automated yet, that'll be done in #47026 but for now please just manually edit that file.

jason810496 replied:
Sure, I will be available to fix about in a hour.

jason810496 replied:
I searched for `external_trigger` in `task_sdk/src/airflow/sdk/api/datamodels/_generated.py` and it looks like I have already removed it.

https://github.com/apache/airflow/pull/45961/files#diff-faa54b815ea1397c8bf23ba8017a4d39e01ae0d2134ac36d40cfebb6e9bdfb0fL302

ashb reviewed (commented):
LGTM, one small change needed in task_sdk as a result though please.
